### PR TITLE
Support experimental feature pragma

### PR DESCRIPTION
**Pull request by axic**
Implements #2346.

### CODE REVIEW

**Comment by axic** [comment]
This has no tests yet, let us first agree it does everything the way it should:
- `pragma` must be followed by an identifier (i.e. cannot be any other token), which can be followed by any token
- experimental must be followed by at least one identifier, but multiple are accepted
- duplicates are forbidden

Now when checking for a specific feature, I'd say to check if `experimentalFeatures.count(feature) || experimentalFeatures.count('*')`. We could add a helper for that on the annotation too: `bool enablesExperimentalFeature(key)`.

**Comment by axic** [comment]
Possibly also it should issue a warning whenever the pragma is encountered.

**Comment by roadriverrail** [comment]
I think this is a good start.  Right now, it looks like if I typed in a non-existant experimental feature (e.g. "pragma experimental foo"), this would end up being basically a no-op, so we don't really get the benefit of the compiler spotting typos.  What if experimental features could register their pragma names, and then if they weren't on the list, the `SyntaxChecker` could issue an error)?

**Comment by axic** [comment]
I think to keep it simple just add the supported const list in `SyntaxChecker`.

**Comment by axic** [comment]
Also the wildcard will not be able to do much type checking :)

**Comment by roadriverrail** [comment]
I guess since we currently only have two features, a simple const list in the SyntaxChecker works.  No reason to get overly complicated.

**Comment by axic** [comment]
Also once any of these features (z3, overflow check) graduate from experimental status, we could still keep them optionally turned on by `pragma feature z3;`

File: libsolidity/analysis/SyntaxChecker.cpp [hunk 1]
```diff
@@ -57,15 +58,44 @@ void SyntaxChecker::endVisit(SourceUnit const& _sourceUnit)
...
 bool SyntaxChecker::visit(PragmaDirective const& _pragma)
 {
 	solAssert(!_pragma.tokens().empty(), "");
 	solAssert(_pragma.tokens().size() == _pragma.literals().size(), "");
-	if (_pragma.tokens()[0] != Token::Identifier || _pragma.literals()[0] != "solidity")
-		m_errorReporter.syntaxError(_pragma.location(), "Unknown pragma \"" + _pragma.literals()[0] + "\"");
-	else
+	if (_pragma.tokens()[0] != Token::Identifier)
+		m_errorReporter.syntaxError(_pragma.location(), "Invalid pragma \"" + _pragma.literals()[0] + "\"");
+	else if (_pragma.literals()[0] == "experimental")
```
Comment (chriseth) [inline libsolidity/analysis/SyntaxChecker.cpp:71]: Depending on the feature, this has to be done much earlier, probably best in the parser.

Comment (chriseth) [reply libsolidity/analysis/SyntaxChecker.cpp:71]: Sorry, confused that with SemanticAnalyzer.

**Comment by axic** [comment]
Idea: how about adding an experimental tag into the metadata CBOR if any experimental mode is selected?

Reason: expose that the contract should not be relied on in the bytecode and not only in the metadata JSON, which very linkely will not be uploaded anywhere.

Effect: regular contract validators will just deny because the CBOR looks invalid to them, and those which understand the flag will have to stop because of the flag.

Example:
```
{bzzr:"<swarmhash>",experimental:true}

File: libsolidity/analysis/SyntaxChecker.cpp (hunk 1 above)
Comment (axic) [reply libsolidity/analysis/SyntaxChecker.cpp:71]: Agreed, but that requires a few more helpers as the parser doesn't currently maintain what sourceunit it is in.

**Comment by axic** [comment]
We've also decided offline to:
- have one pragma experimental per feature
- do not support wildcard
- have an enum for the features
- keep the enum/string mapping in a new header `AST/ExperimentalFeatures.h`

**Comment by axic** [comment]
I'd like to split off the metadata stamp changes to a separate PR, because that might require more decision, while this above should be ready.

**Comment by axic** [comment]
@chriseth do you want to review this after the release? Perhaps we could reduce the number of error messages, it may be too much detail.

File: libsolidity/ast/ExperimentalFeatures.h [hunk 2]
```diff
@@ -0,0 +1,35 @@
...
+#include <map>
+
+namespace dev
+{
+namespace solidity
+{
+
+enum class ExperimentalFeature {};
+
+static const std::map<std::string, ExperimentalFeature> experimentalFeatureNames = {};
```
Comment (chriseth) [inline libsolidity/ast/ExperimentalFeatures.h]: Can you call this `g_experimentalFeatureNames` or put it inside a class?

**Comment by axic** [comment]
Updated `ExperimentalFeatureNames` to start with an uppercase letter as it is a global (similar to Version.h)

**Review by chriseth — APPROVED** [review]

**Comment by duaraghav8** [comment]
@axic  slightly confused here. Are multiple identifiers allowed after `experimental`? If yes, what's the syntax? (Tried comma-separated & space-separated with no luck)

**Comment by axic** [comment]
No, it was changed to only a single one.

**Pull request merged**